### PR TITLE
Improve the togglecontrol documentation

### LINKED ISSUE #23740: ToggleControl Documentation
The `ToggleControl` documentation uses state exclusively to demonstrate something that should be a block attribute or provided as a prop from elsewhere.

As a result I just spent a while trying to undo some misconceptions on stack exchange by a very confused person who thought the way to put a toggle control in a block was to use `withState`, and didn't know how to retrieve that state in the blocks save function. https://wordpress.stackexchange.com/questions/370550/gutenberg-get-withstate-variable-in-save-function

Instead the example should be replaced with a more comprehensive/clearer version, ideally using a block as an example.

I understand in this case `withState` will have been used because it allowed for a smaller example under the premise that this would help developers. However, this has made the example actively misleading, and raised the barrier to entry to require senior developer React skills.

### PR DESCRIPTION
**Pull request by tomjn**
## Description

Closes #23740 by documenting how to use the control inside a block

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/toggle-control/README.md [hunk 1]
```diff
@@ -22,6 +25,38 @@ const MyToggleControl = withState( {
 ) );
 ```
 
+To use this inside a block to toggle a saveable value, use block attributes, e.g.
```
Comment (youknowriad) [inline packages/components/src/toggle-control/README.md:28]: While I think this is valuable, I'm not sure it's the right place for it. Why just the toggle control, why not all *controls?. I think the components documentation should probably stay agnostic to the context where the component is used.

Comment (tomjn) [reply packages/components/src/toggle-control/README.md:28]: 1. People don't know much about gutenberg so they look up the individual components for how to use them in their blocks
2. Examples are seen as canonical. You and I might understand that `withState` is a tool being used here, but the general consensus to those not in the know is that it's a requirement for using toggle control components.
3. I have a busy working day, doing this for every component is a nice idea, it also sounds like a full day of work, I've a deadline tomorrow so my project manager might want to have words :)

If we want to make the component documentation agnostic then removing the `withState` and having placeholder variables such as:

```js
const setToggleValue = ...
const currentValue = ...;
```

Otherwise the context agnostic approach is not being followed here, and, is actively misleading. I understand there's a certain appeal to keeping components in their pristine context-less state, but people need examples. Context in and of itself can be useful.

If the case is that all documentation should be consistent, then I agree, but that shouldn't prevent fixes to technical documentation. "Either all components can have it or none of them can have it" basically means none of them will have it.

Comment (tomjn) [reply packages/components/src/toggle-control/README.md:28]: And if this isn't the place, then where is? And should all these components not link there?

This isn't the first time I've seen components documented in isolation causing issues. Important details are left out because they're not directly a part of the component, but rather their dependencies.

Comment (youknowriad) [reply packages/components/src/toggle-control/README.md:28]: It's not about writing less documentation, it's because it doesn't feel like the right place for it for me at least. Maybe it's about documentation organization.

For me you want to write a block:

 - Learn React that way you learn how components work (so we potentially need links to React docs) and it's not necessary to repeat the same documentation over and over for each component.
 - Learn the Block API and learn that you can use any React component there including the provided ones. Link to the components docs there with potentially an example of how to use one of these.

So I see several things:

 - Do we have links to React docs anywhere? 
 - Do we have links to components from the "create blocks tutorials"?

To put it differently (and also a bit controversially :)) Do we explain how to use PHP Classes and PHP function in each WordPress PHP API? 

And potentially to get the best of both world, link to React docs at the top of each component README?

**Comment by tomjn** [comment]
Actually it’s seen as poor practice in a lot of situations to mess with the
internals of the PHP APIs objects. Their documentation is also severely
lacking. They also don’t have methods or take parameters and act as return
values rather than components. WP_Query being the notable exception, for
which there are lots of examples. Otherwise there are very few equivalent
examples in the PHP side of things to what we have here.

I would not use them as an example of what to do, most of them are just the
classes source code presented in a collapsed syntax highlighted box.

In the meantime a lot of people are using react for the first time and
using it with these components. We can’t ask them to go away on a course
beforehand. They may even have done a course and are unsure.

In this case they open the documentation for this component and see that
the canonical usage of this component involves a function named withState
that isn’t explained. I tried to google for it and found no results and had
to use github search.

Given that each component is missing all context in their documentation it
would be reasonable to assume that withState is a required part of using
this component.

**Comment by youknowriad** [comment]
@tomjn I understand your point and I agree that `withState` should be removed from the components docs. that's from a time where React's default `useState` didn't exist and should be easier to understand for everyone. Now for the documentation itself, I still think the component READMEs are not the right place for this: @wordpress/components is not meant to be used in blocks only, it's a generic react component library, so its documentation should be optimized as such IMO. It can be used in blocks, in WordPress screens, outside WordPress context...

**Comment by youknowriad** [comment]
Added the documentation label to have more eyes on this. I'm not trying to force any way or the other, I'm just convinced that it's the right place for it and open to alternative opinions if folks think otherwise.

**Comment by tomjn** [comment]
> @wordpress/components is not meant to be used in blocks only

I didn't suggest it was, if the concern is that an example using a block implies these components can only be used in blocks, then I would understand if the original example had been removed, but it wasn't.

Instead, a selection of examples demonstrating the use of the component in different scenarios would highlight how to use it in a block and demonstrate that it isn't unique to blocks. Additional examples reinforces that the component is its own item.

So instead, demonstrate that it's agnostic by showing it in lots of contexts, rather than being ultra-minimal. An examples section containing several examples under titles makes sense. As long as everything is clearly marked it should be fine.

I'll update the PR to reflect that approach.

I also think that it makes more sense to say this is a standalone component. Saying it's agnostic isn't easy to understand and it's ambiguous..

**Comment by tomjn** [comment]
I've changed this with an example section, and rewritten the original example to use the `useState` React hook. I also made some wording adjustments

**Comment by chrisdavies71** [comment]
@tomjn Thanks for the work on this. I was having exactly the issue you described and thought I was going mad. Your way of laying out the code examples works really well as it makes things easy to understand.

**Comment by fabiankaegy** [comment]
Given that the docs have since been updated to use `useState` I'm going to close this PR for now 👍 Feel free to reopen if you think there is still something you want to update further.
